src/mon/NVMeofGwMon.cc

```diff
@@ -0,0 +1,497 @@
+
+
+    //handle exception of tick overdued in order to avoid false detection of overdued beacons , see MgrMonitor::tick
+    const auto cutoff = now - nvmegw_beacon_grace;
+    for(auto &itr : last_beacon){// Pass over all the stored beacons
+        auto& lb = itr.first;
+        auto last_beacon_time = itr.second;
+        if(last_beacon_time < cutoff){
+            dout(4) << "beacon timeout for GW " << lb.gw_id << dendl;
+            pending_map.process_gw_map_gw_down( lb.gw_id, lb.group_key, propose);
+            _propose_pending |= propose;
+            last_beacon.erase(lb);
+        }
+        else {
+           dout(20) << "beacon live for GW key: " << lb.gw_id << dendl;
+        }
+    }
+
+    pending_map.handle_abandoned_ana_groups(propose); // Periodic: take care of not handled ANA groups
+    _propose_pending |= propose;
+
+    if(_propose_pending){
+       //pending_map.delay_propose = true; // not to send map to clients immediately in "update_from_paxos"
+       dout(4) << "decision to delayed_map" <<dendl;
+       propose_pending();
+    }
+
+    // if propose_pending returned true , call propose_pending method of the paxosService
+    // todo understand the logic of paxos.plugged for sending several propose_pending see MgrMonitor::tick
+}
+
+const char **NVMeofGwMon::get_tracked_conf_keys() const
+{
+  static const char* KEYS[] = {
+    NULL
+  };
+  return KEYS;
+}
+
+void NVMeofGwMon::handle_conf_change(const ConfigProxy& conf,
+                                    const std::set<std::string> &changed)
+{
+  dout(4) << "changed " << changed << dendl;
+}
+
+void NVMeofGwMon::create_pending(){
+
+    pending_map = map;// deep copy of the object
+    // TODO  since "pending_map"  can be reset  each time during paxos re-election even in the middle of the changes ...
+    pending_map.epoch++;
+    dout(4) << " pending " << pending_map  << dendl;
+    if(last_leader == false){ // peon becomes leader and gets updated map , need to synchronize the last_beacon
+        synchronize_last_beacon();
+        last_leader = true;
+    }
+}
+
+void NVMeofGwMon::encode_pending(MonitorDBStore::TransactionRef t){
+
+    dout(10) <<  dendl;
+    ceph_assert(get_last_committed() + 1 == pending_map.epoch);
+    bufferlist bl;
+    pending_map.encode(bl);
+    put_version(t, pending_map.epoch, bl);
+    put_last_committed(t, pending_map.epoch);
+}
+
+void NVMeofGwMon::update_from_paxos(bool *need_bootstrap){
+    version_t version = get_last_committed();
+
+    //dout(4) <<  MY_MON_PREFFIX << __func__ << " version "  << version  << " map.epoch " << map.epoch << dendl;
+
+    if (version != map.epoch) {
+        dout(4) << " NVMeGW loading version " << version  << " " << map.epoch << dendl;
+
+        bufferlist bl;
+        int err = get_version(version, bl);
+        ceph_assert(err == 0);
+
+        auto p = bl.cbegin();
+        map.decode(p);
+        if(!mon.is_leader()) {
+            dout(4) << "leader map: " << map <<  dendl;
+        }
```

Comment:
```suggestion
            dout(5) << "map: " << map <<  dendl;
```

Would be appropriate I think.

Reply:
I just wanted to  dump the decision map of leader on the peons . The pending_map is already dumped by leader upon solution so this would be like duplicated dump on the leader